### ISSUE TITLE
Design how the various search and filtering modes interact

### ISSUE THREAD

**Issue by robertknight** [hypothesis/client#6006]
**_Update 2024-01-17_**: See https://github.com/hypothesis/client/issues/6006#issuecomment-1883077311 for current mocks. 

## Implementation sub-tasks

- [x] Compute accurate counts for each tab when a search or filter is active
- [x] Display tabs when a search or filter is active
- [x] Update the UI layout of search/filter results to move them into the same panel (https://github.com/hypothesis/client/pull/6106)
- [x] Display the search/filter panel when no search is active, but a filter is configured (https://github.com/hypothesis/client/pull/6108)
- [x] Update design of filter toggles to match the current mocks (https://github.com/hypothesis/client/pull/6106)
- [x] Use minus rather than plus icon in filter toggle buttons when active (https://github.com/hypothesis/client/pull/6107)
- [x] Resolve issue with empty filters box when grading (https://github.com/hypothesis/client/issues/6109)
- [x] Update design of search bar to match the mocks (https://github.com/hypothesis/client/pull/6130)
- [x] Figure out how we're going to handle annotation selections (when annotations are clicked in the document) in the new design (https://github.com/hypothesis/client/pull/6117, https://github.com/hypothesis/client/pull/6118)
- [x] Make sure appropriate screen reader status updates are in place as search results change (https://github.com/hypothesis/client/pull/6128)
- [x] Test end-to-end workflow when using page ranges in VS assignments to filter annotations to selected part of book

----

Below is the original problem outline and discussion about how to update the design.

The Hypothesis client has several annotation filtering mechanisms, some of which can be active concurrently. This issue is about clarifying how those mechanisms should interact.

## Available filtering mechanisms

The different filtering mechanisms that exist in the Hypothesis client are:

1. **Search field**. This is the input field revealed by search icon in the top bar of the client.
2. **Selection**. This is created when a user clicks on a highlight in the document, or a user follows a direct to an annotation.
3. **User focus**. This is enabled when a teacher is grading a specific student's annotations in the LMS app
4. **Content focus** (NEW). This is a planned new focus mode that would work similar to user focus. It would be enabled in the LMS app when launching an assignment that has been configured to use only part of a document (eg. a specific chapter or page range from a book or PDF). See https://github.com/hypothesis/client/issues/5937.

In addition to the above filtering modes, there are conditions under which an annotation that does not match the filters can still be shown. These are:

A. When an annotation is being edited. This occurs when the user creates a new annotation that does not match the current filters, and has not saved it yet.
B. When part of a conversation thread that does not match the filter is revealed by clicking a "Show more in conversation" button

## Interactions between filters

The way the filters currently interact is as follows:

1. The search field and user focus filters are combined. So entering a search while user focus is active, searches within the user's annotations.
2. The selection _replaces_ the search and user focus when active. This means that if you click on a highlight in the document, you'll see the annotation whether it matches the filter or not. This does raise the issue of whether searching in the sidebar ought to change the highlights shown in the page. Currently it doesn't.

## Screenshots of individual filters

**Search field (old):**

<img width="480" alt="search input field" src="https://github.com/hypothesis/client/assets/2458/0c062904-d53f-481e-ba58-4d85bae59384">

**Search field (new):**

The new search UI moves the search bar into a panel below the header, where we avoid the problem of being very space constrained, leading to issues like https://github.com/hypothesis/client/issues/5963.

<img width="481" alt="Search filter - new" src="https://github.com/hypothesis/client/assets/2458/09022765-602a-4e9f-af82-a1ef0346c67a">

**Selection:**

After clicking a highlight in the document, you get this:

<img width="487" alt="Selection active" src="https://github.com/hypothesis/client/assets/2458/952ec93a-35f1-466c-bcfd-5deb255aba10">

Note there is no indication here about _why_ the annotation is being shown.

**User focus (active):**

A teacher's view when grading a student:

<img width="643" alt="Grading view user focus" src="https://github.com/hypothesis/client/assets/2458/473c6cc4-588a-42d2-91f6-2a97419333ae">

**User focus (inactive):**

If the user clicks on the "Show all" button, the above changes to:

<img width="640" alt="Grading view user focus inactive" src="https://github.com/hypothesis/client/assets/2458/8def7ef7-a02f-467b-b440-b7d003d2d6e8">

**Content focus:**

The planned focus more for content would be similar to the above focus mode for users, but instead of a user's display name, you'd see "pages 100-120", or whatever the selected content range for the assignment is.

## Screenshots of combined filters

**Search field (old) + user focus:**

<img width="471" alt="Filter and focus" src="https://github.com/hypothesis/client/assets/2458/7f54d524-92f4-4c32-b01b-d5859dc083f4">

This is compact but user confusion can arise from the fact there are two filters active, but only one button to clear them. Clicking the "Clear search" button both **_clears the search AND disables the focus mode_**. This means that when user focus mode is active, if you perform a search you need to perform multiple actions to get back to the previous state (user focus active, no search).

**Search field (new) + user focus:**

<img width="480" alt="Filter and focus active" src="https://github.com/hypothesis/client/assets/2458/4209ea4c-23f5-4312-a7b6-0649354ada90">

This UI takes up more space, but makes it clearer that there are two distinct filters active and it should be possible to clear only one of them at a time (TODO: Confirm that the cancel button actually does do this).

## Questions to address

----

1. How should the different modes interact with one another? (eg. intersection of filters as we have with search + user focus, vs replacement as we have with selection replacing search and/or focus)
2. How should we display the combined modes?
3. When you have both a search and focus mode active, how do you clear just one of them?
4. When you make a selection, can we make it possible to clear that without losing the previous search state? 
5. If we introduce a new "focus on a page range" mode, how does that get displayed?

**Comment by robertknight** [comment]
Some relevant UI from other apps that provides some kind of "filter a list with multiple active filters" UI.

**Slack:**

<img width="361" alt="Slack conversation filter" src="https://github.com/hypothesis/client/assets/2458/6dcb1e6d-58e2-48ad-9c0d-76d8d1cd7056">

If for example you check "Without external people" you see this at the top:

<img width="422" alt="Without external people" src="https://github.com/hypothesis/client/assets/2458/8ff49293-3b0c-4d74-9cb8-a842143bfe78">

**Comment by robertknight** [comment]
Something I don't like about the current filter toggle UI is that the button can dramatically change length depending on whether the filter is active or not. This means that the text moves around in a distracting way when it is toggled, and also the hit target changes. So a fixed sized icon is a must in the replacement UI.

**Comment by robertknight** [comment]
Jared, Dan, Alejandro and myself discussed this on a call yesterday. We arrived at the designs discussed in https://hypothes-is.slack.com/archives/C07NXBDNW/p1704730651476379 (Figma: https://www.figma.com/file/sC8gPsiPLHOuaGKSCvQXUR/Hypothesis---Search-(Sidebar)?type=design&node-id=0-1&mode=design&t=94q4pMDVYxpeYYTi-0):

![Hypothesis - Search (Sidebar)](https://github.com/hypothesis/client/assets/2458/bf6506f0-98ee-4fef-aa44-d516a2c6de83)

Some notes:

- The current single focus filter on/off toggle is replaced with individual buttons per focus state, as in https://github.com/hypothesis/client/pull/6070
- The focus toggle buttons are integrated into the filter panel. If there is not enough room to fit all the toggles, they wrap.
- Tabs are shown whether a search is active or not
- The label reporting the number of matches has been removed, since that number is displayed on the tabs (albeit divided into Annotations and Page Notes)
- When there are focus filters configured, the filter panel is always visible, but does not include the search bar if the search toggle is not pressed.
- The "X" action for the panel is removed. Instead an "X" is shown inside the search bar. The closest existing UI we have for this is the button to clear the transcript search in Via's video annotation UI.
